transformer-js requires GLIBC 2.29

# 🐛 bug report
I'm trying to get Parcel 2 to run on an arm64 device running Ubuntu 18.04.5 LTS. `parcel serve` triggers a build which terminates with the error

```
Error: /lib/aarch64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /home/[...]/node_modules/@parcel/transformer-js/parcel-swc.linux-arm64-gnu.node)
```

Ubuntu 18.04 only has glibc 2.27 so perhaps I just need to get a newer version of glibc on the system – but looking at previous issues, I get the impression that Parcel is not supposed to require that, thus this issue.

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.0.0-rc.0 |
| Node             | v16.4.1 |
| npm/Yarn         | npm 7.20.6 |
| Operating System | Ubuntu 18.04.5 LTS |

> @JaedsonBarbosa I don't know if it helps in your case but when I build Parcel from source, a working version of `parcel-swc.linux-arm64-gnu.node` is built for me.

I am currently getting this error message:
`Error: Cannot find module './artifacts/index.linux-arm64-gnu.node'`

> I am currently getting this error message:
> `Error: Cannot find module './artifacts/index.linux-arm64-gnu.node'`

@JaedsonBarbosa Are you sure that's related to `transformer-js` and not `source-map`?

OT but if it is the latter then a workaround is to add `@parcel/source-map@2.0.0-rc.6` to your project. Then install `@parcel/source-map@2.0.0-rc.7` somewhere else and move the binary from the artefacts directory there to the rc6 directory. I.e.
```
npm init -y
npm install @parcel/source-map@2.0.0-rc.7
cp node_modules/@parcel/source-map/parcel_sourcemap_node/artifacts/index.linux-arm64-gnu.node [your project path]/node_modules/@parcel/source-map/parcel_sourcemap_node/artifacts
```

--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -1,8 +1,5 @@
 [target.'cfg(target_env = "gnu")']
 rustflags = ["-C", "link-args=-Wl,-z,nodelete"]
 
-[target.aarch64-unknown-linux-musl]
-linker = "aarch64-linux-musl-gcc"
-
 [target.wasm32-unknown-unknown]
 rustflags = ["-C", "link-arg=--export-table"]

--- a/.cargo/config.toml
+++ b/.cargo/config.toml
@@ -1,18 +1,6 @@
 [target.'cfg(target_env = "gnu")']
 rustflags = ["-C", "link-args=-Wl,-z,nodelete"]
 
-[target.arm-unknown-linux-gnueabihf]
-linker = "arm-linux-gnueabihf-gcc"
-rustflags = ["-C", "link-args=-Wl,-z,nodelete"]
-
-[target.armv7-unknown-linux-gnueabihf]
-linker = "arm-linux-gnueabihf-gcc"
-rustflags = ["-C", "link-args=-Wl,-z,nodelete"]
-
-[target.aarch64-unknown-linux-gnu]
-linker = "aarch64-linux-gnu-gcc"
-rustflags = ["-C", "link-args=-Wl,-z,nodelete"]
-
 [target.aarch64-unknown-linux-musl]
 linker = "aarch64-linux-musl-gcc"
 

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -81,33 +81,33 @@ jobs:
         include:
           - target: x86_64-unknown-linux-gnu
             arch: x86_64
-            cflags: -mmevex512
+            cflags: -mevex512
             glibc: 2.28
-            docker: node:18-slim
+            docker: node:20-slim
             args: ''
           - target: armv7-unknown-linux-gnueabihf
             arch: armhf
             cflags: -mfpu=neon
             glibc: 2.28
-            docker: node:18-slim
+            docker: node:20-slim
             args: '--platform linux/arm/v7'
           - target: aarch64-unknown-linux-gnu
             arch: arm64
             cflags: ''
             glibc: 2.28
-            docker: node:18-slim
+            docker: node:20-slim
             args: '--platform linux/arm64'
           - target: x86_64-unknown-linux-musl
             arch: x86_64
             cflags: -msse4.2 -mevex512
             glibc: ''
-            docker: node:18-alpine
+            docker: node:20-alpine
             args: ''
           - target: aarch64-unknown-linux-musl
             arch: aarch64
             cflags: ''
             glibc: ''
-            docker: node:18-alpine
+            docker: node:20-alpine
             args: '--platform linux/arm64'
     name: ${{ matrix.target }}
     runs-on: ubuntu-latest
@@ -155,15 +155,11 @@ jobs:
         if: ${{ matrix.arch != 'x86_64' }}
         run: docker run --rm --privileged multiarch/qemu-user-static:register --reset
       - name: Smoke test
-        if: ${{ matrix.arch != 'x86_64' }}
         uses: addnab/docker-run-action@v3
         with:
           image: ${{ matrix.docker }}
           options: ${{ matrix.args }} -v ${{ github.workspace }}:/build -w /build
           run: node -e "require('@parcel/rust')"
-      - name: Smoke test
-        if: ${{ matrix.arch == 'x86_64' }}
-        run: node -e 'require("@parcel/rust")'
 
   build-and-release:
     runs-on: ubuntu-22.04

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -74,45 +74,43 @@ jobs:
         if: ${{ !matrix.target }}
         run: node -e "require('@parcel/rust')"
 
-  build-linux-gnu-x64:
-    name: linux-gnu-x64
-    runs-on: ubuntu-22.04
-    container:
-      image: node:22.4-bullseye
-    steps:
-      - uses: actions/checkout@v3
-      - name: Install Rust
-        uses: actions-rs/toolchain@v1
-        with:
-          toolchain: stable
-          profile: minimal
-          override: true
-      - uses: bahmutov/npm-install@v1.8.35
-      - name: Build native packages
-        run: yarn build-native-${{ inputs.profile }}
-      - name: Upload artifacts
-        uses: actions/upload-artifact@v4
-        with:
-          name: bindings-linux-gnu-x64
-          path: packages/*/*/*.node
-      - name: debug
-        run: ls -l packages/*/*/*.node
-      - name: Smoke test
-        run: node -e 'require("@parcel/rust")'
-
-  build-linux-gnu-arm:
+  build-linux:
     strategy:
       fail-fast: false
       matrix:
         include:
-          - target: arm-unknown-linux-gnueabihf
+          - target: x86_64-unknown-linux-gnu
+            arch: x86_64
+            cflags: ''
+            glibc: 2.28
+            docker: node:18-slim
+            args: ''
+          - target: armv7-unknown-linux-gnueabihf
             arch: armhf
             cflags: -mfpu=neon
+            glibc: 2.28
+            docker: node:18-slim
+            args: '--platform linux/arm/v7'
           - target: aarch64-unknown-linux-gnu
             arch: arm64
             cflags: ''
+            glibc: 2.28
+            docker: node:18-slim
+            args: '--platform linux/arm64'
+          - target: x86_64-unknown-linux-musl
+            arch: x86_64
+            cflags: -msse4.2
+            glibc: ''
+            docker: node:18-alpine
+            args: ''
+          - target: aarch64-unknown-linux-musl
+            arch: aarch64
+            cflags: ''
+            glibc: ''
+            docker: node:18-alpine
+            args: '--platform linux/arm64'
     name: ${{ matrix.target }}
-    runs-on: ubuntu-22.04
+    runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
       - name: Install Rust
@@ -139,7 +137,7 @@ jobs:
         run: yarn build-native-${{ inputs.profile }}
         env:
           RUST_TARGET: ${{ matrix.target }}
-          ZIG_GLIBC: 2.28
+          ZIG_GLIBC: ${{ matrix.glibc }}
           CFLAGS: ${{ matrix.cflags }}
       - name: Upload artifacts
         uses: actions/upload-artifact@v4
@@ -148,66 +146,30 @@ jobs:
           path: packages/*/*/*.node
       - name: debug
         run: ls -l packages/*/*/*.node
+      - name: Set up QEMU
+        if: ${{ matrix.arch != 'x86_64' }}
+        uses: docker/setup-qemu-action@v3
+        with:
+          platforms: arm64,arm
       - name: Configure binfmt-support
+        if: ${{ matrix.arch != 'x86_64' }}
         run: docker run --rm --privileged multiarch/qemu-user-static:register --reset
       - name: Smoke test
-        uses: addnab/docker-run-action@v1
+        if: ${{ matrix.arch != 'x86_64' }}
+        uses: addnab/docker-run-action@v3
         with:
-          image: ghcr.io/devongovett/multiarch-node:node14-${{ matrix.arch }}-focal
-          options: -v ${{github.workspace}}:/work
+          image: ${{ matrix.docker }}
+          options: ${{ matrix.args }} -v ${{ github.workspace }}:/work -w /work
           run: cd /work && node -e "require('@parcel/rust')"
-
-  build-linux-musl:
-    strategy:
-      fail-fast: false
-      matrix:
-        include:
-          - target: x86_64-unknown-linux-musl
-            cflags: -msse4.2
-            arch: x86_64
-          - target: aarch64-unknown-linux-musl
-            cflags: ''
-            arch: aarch64
-    name: ${{ matrix.target }}
-    runs-on: ubuntu-22.04
-    container:
-      image: ghcr.io/napi-rs/napi-rs/nodejs-rust@sha256:5a99e45446355d25c20e95d35231d84e9ce472280d8c0b1be53281bade905f09
-      credentials:
-        username: ${{ github.actor }}
-        password: ${{ secrets.GHCR_TOKEN }}
-    steps:
-      - uses: actions/checkout@v4
-      - name: Install build tools
-        run: apk add --no-cache python3 make gcc g++ musl-dev curl openssl
-      - name: Install Rust
-        uses: dtolnay/rust-toolchain@master
-        with:
-          toolchain: stable
-          target: ${{ matrix.target }}
-      - uses: bahmutov/npm-install@v1.8.35
-      - name: Build native packages
-        run: yarn build-native-${{ inputs.profile }}
-        env:
-          RUST_TARGET: ${{ matrix.target }}
-          CFLAGS: ${{ matrix.cflags }}
-      - name: Upload artifacts
-        uses: actions/upload-artifact@v4
-        with:
-          name: bindings-linux-musl-${{ matrix.arch }}
-          path: packages/*/*/*.node
-      - name: debug
-        run: ls -l packages/*/*/*.node
       - name: Smoke test
-        if: ${{ matrix.target == 'x86_64-unknown-linux-musl' }}
+        if: ${{ matrix.arch == 'x86_64' }}
         run: node -e 'require("@parcel/rust")'
 
   build-and-release:
     runs-on: ubuntu-22.04
     name: Build and release
     needs:
-      - build-linux-gnu-arm
-      - build-linux-gnu-x64
-      - build-linux-musl
+      - build-linux
       - build-macos-windows
     steps:
       - uses: actions/checkout@v4

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -138,7 +138,8 @@ jobs:
       - name: Build native packages
         run: yarn build-native-${{ inputs.profile }}
         env:
-          RUST_TARGET: ${{ matrix.target }}.2.28
+          RUST_TARGET: ${{ matrix.target }}
+          ZIG_GLIBC: 2.28
           CFLAGS: ${{ matrix.cflags }}
       - name: Upload artifacts
         uses: actions/upload-artifact@v4

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -105,10 +105,10 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - target: arm-unknown-linux-gnueabihf.2.28
+          - target: arm-unknown-linux-gnueabihf
             arch: armhf
             cflags: -mfpu=neon
-          - target: aarch64-unknown-linux-gnu.2.28
+          - target: aarch64-unknown-linux-gnu
             arch: arm64
             cflags: ''
     name: ${{ matrix.target }}
@@ -138,7 +138,7 @@ jobs:
       - name: Build native packages
         run: yarn build-native-${{ inputs.profile }}
         env:
-          RUST_TARGET: ${{ matrix.target }}
+          RUST_TARGET: ${{ matrix.target }}.2.28
           CFLAGS: ${{ matrix.cflags }}
       - name: Upload artifacts
         uses: actions/upload-artifact@v4

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -81,7 +81,7 @@ jobs:
         include:
           - target: x86_64-unknown-linux-gnu
             arch: x86_64
-            cflags: ''
+            cflags: -mmevex512
             glibc: 2.28
             docker: node:18-slim
             args: ''
@@ -99,7 +99,7 @@ jobs:
             args: '--platform linux/arm64'
           - target: x86_64-unknown-linux-musl
             arch: x86_64
-            cflags: -msse4.2
+            cflags: -msse4.2 -mevex512
             glibc: ''
             docker: node:18-alpine
             args: ''
@@ -159,8 +159,8 @@ jobs:
         uses: addnab/docker-run-action@v3
         with:
           image: ${{ matrix.docker }}
-          options: ${{ matrix.args }} -v ${{ github.workspace }}:/work -w /work
-          run: cd /work && node -e "require('@parcel/rust')"
+          options: ${{ matrix.args }} -v ${{ github.workspace }}:/build -w /build
+          run: node -e "require('@parcel/rust')"
       - name: Smoke test
         if: ${{ matrix.arch == 'x86_64' }}
         run: node -e 'require("@parcel/rust")'

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -153,7 +153,7 @@ jobs:
           platforms: arm64,arm
       - name: Configure binfmt-support
         if: ${{ matrix.arch != 'x86_64' }}
-        run: docker run --rm --privileged multiarch/qemu-user-static:register --reset
+        run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
       - name: Smoke test
         uses: addnab/docker-run-action@v3
         with:

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -62,18 +62,6 @@ jobs:
         run: yarn build-native-${{ inputs.profile }}
         env:
           RUST_TARGET: ${{ matrix.target }}
-      - name: Extract debug symbols
-        if: ${{ runner.os == 'macOS' && inputs.profile == 'canary' }}
-        run: dsymutil packages/*/*/*.node
-      - name: Upload debug symbols
-        uses: actions/upload-artifact@v4
-        if: ${{ inputs.profile == 'canary' }}
-        with:
-          name: debug-symbols-${{ matrix.name }}
-          path: packages/*/*/*.node.dSYM/Contents/Resources/DWARF/*.node
-      - name: Strip debug symbols # https://github.com/rust-lang/rust/issues/46034
-        if: ${{ runner.os == 'macOS' }}
-        run: strip -x packages/*/*/*.node # Must use -x on macOS. This produces larger results on linux.
       - name: Upload artifacts
         uses: actions/upload-artifact@v4
         with:
@@ -102,20 +90,6 @@ jobs:
       - uses: bahmutov/npm-install@v1.8.35
       - name: Build native packages
         run: yarn build-native-${{ inputs.profile }}
-      - name: Extract debug symbols
-        if: ${{ inputs.profile == 'canary' }}
-        run: |
-          find packages -name "*.node" -type f -exec objcopy --only-keep-debug --compress-debug-sections=zlib {} {}.debug \;
-          find packages -name "*.node" -type f -exec objcopy --strip-debug --strip-unneeded {} \;
-          find packages -name "*.node" -type f -exec objcopy --add-gnu-debuglink={}.debug {} \;
-      - name: Upload debug symbols
-        uses: actions/upload-artifact@v4
-        if: ${{ inputs.profile == 'canary' }}
-        with:
-          name: debug-symbols-linux-gnu-x64
-          path: packages/*/*/*.node.debug
-      - name: Strip debug symbols # https://github.com/rust-lang/rust/issues/46034
-        run: strip packages/*/*/*.node
       - name: Upload artifacts
         uses: actions/upload-artifact@v4
         with:
@@ -131,15 +105,11 @@ jobs:
       fail-fast: false
       matrix:
         include:
-          - target: arm-unknown-linux-gnueabihf
+          - target: arm-unknown-linux-gnueabihf.2.28
             arch: armhf
-            strip: arm-linux-gnueabihf-strip
-            objcopy: arm-linux-gnueabihf-objcopy
             cflags: -mfpu=neon
-          - target: aarch64-unknown-linux-gnu
+          - target: aarch64-unknown-linux-gnu.2.28
             arch: arm64
-            strip: aarch64-linux-gnu-strip
-            objcopy: aarch64-linux-gnu-objcopy
             cflags: ''
     name: ${{ matrix.target }}
     runs-on: ubuntu-22.04
@@ -150,10 +120,16 @@ jobs:
         with:
           toolchain: stable
           target: ${{ matrix.target }}
-      - name: Install cross compile toolchains
-        run: |
-          sudo apt-get update
-          sudo apt-get install gcc-arm-linux-gnueabihf gcc-aarch64-linux-gnu g++-aarch64-linux-gnu -y
+      - name: Install ziglang
+        uses: goto-bus-stop/setup-zig@v1
+        with:
+          version: 0.13.0
+      - name: Install cargo toolchains
+        uses: taiki-e/install-action@v2
+        env:
+          GITHUB_TOKEN: ${{ github.token }}
+        with:
+          tool: cargo-zigbuild
       - uses: bahmutov/npm-install@v1.8.35
       - uses: Swatinem/rust-cache@v2
         if: ${{ inputs.type != 'latest' }}
@@ -164,20 +140,6 @@ jobs:
         env:
           RUST_TARGET: ${{ matrix.target }}
           CFLAGS: ${{ matrix.cflags }}
-      - name: Extract debug symbols
-        if: ${{ inputs.profile == 'canary' }}
-        run: |
-          find packages -name "*.node" -type f -exec ${{ matrix.objcopy }} --only-keep-debug --compress-debug-sections=zlib {} {}.debug \;
-          find packages -name "*.node" -type f -exec ${{ matrix.objcopy }} --strip-debug --strip-unneeded {} \;
-          find packages -name "*.node" -type f -exec ${{ matrix.objcopy }} --add-gnu-debuglink={}.debug {} \;
-      - name: Upload debug symbols
-        uses: actions/upload-artifact@v4
-        if: ${{ inputs.profile == 'canary' }}
-        with:
-          name: debug-symbols-${{ matrix.target }}
-          path: packages/*/*/*.node.debug
-      - name: Strip debug symbols # https://github.com/rust-lang/rust/issues/46034
-        run: ${{ matrix.strip }} packages/*/*/*.node
       - name: Upload artifacts
         uses: actions/upload-artifact@v4
         with:
@@ -200,13 +162,9 @@ jobs:
       matrix:
         include:
           - target: x86_64-unknown-linux-musl
-            strip: strip
-            objcopy: objcopy
             cflags: -msse4.2
             arch: x86_64
           - target: aarch64-unknown-linux-musl
-            strip: aarch64-linux-musl-strip
-            objcopy: aarch64-linux-musl-objcopy
             cflags: ''
             arch: aarch64
     name: ${{ matrix.target }}
@@ -231,20 +189,6 @@ jobs:
         env:
           RUST_TARGET: ${{ matrix.target }}
           CFLAGS: ${{ matrix.cflags }}
-      - name: Extract debug symbols
-        if: ${{ inputs.profile == 'canary' }}
-        run: |
-          find packages -name "*.node" -type f -exec ${{ matrix.objcopy }} --only-keep-debug --compress-debug-sections=zlib {} {}.debug \;
-          find packages -name "*.node" -type f -exec ${{ matrix.objcopy }} --strip-debug --strip-unneeded {} \;
-          find packages -name "*.node" -type f -exec ${{ matrix.objcopy }} --add-gnu-debuglink={}.debug {} \;
-      - name: Upload debug symbols
-        uses: actions/upload-artifact@v4
-        if: ${{ inputs.profile == 'canary' }}
-        with:
-          name: debug-symbols-linux-musl-${{ matrix.arch }}
-          path: packages/*/*/*.node.debug
-      - name: Strip debug symbols # https://github.com/rust-lang/rust/issues/46034
-        run: ${{ matrix.strip }} packages/*/*/*.node
       - name: Upload artifacts
         uses: actions/upload-artifact@v4
         with:
@@ -277,19 +221,6 @@ jobs:
           path: artifacts
       - name: Move bindings
         run: for d in artifacts/bindings-*/*/*; do cp $d/*.node packages/$(basename $(dirname $d))/$(basename $d); done
-      - name: Move debug symbols
-        if: ${{ inputs.profile == 'canary' }}
-        run: |
-          mkdir debug-symbols
-          find artifacts -name "*.debug" -exec cp {} debug-symbols/ \;
-          find artifacts -name "*.node" -path "**/DWARF/**" -exec cp {} debug-symbols/ \;
-          ls -l debug-symbols
-      - name: Upload combined debug symbols artifact
-        uses: actions/upload-artifact@v4
-        if: ${{ inputs.profile == 'canary' }}
-        with:
-          name: debug-symbols
-          path: debug-symbols/**
       - name: Debug
         run: |
           ls -l packages/*/*/*.node

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -123,7 +123,7 @@ jobs:
       - name: Install ziglang
         uses: goto-bus-stop/setup-zig@v1
         with:
-          version: 0.13.0
+          version: 0.14.0
       - name: Install cargo toolchains
         uses: taiki-e/install-action@v2
         env:

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -9,6 +9,9 @@ members = [
 inherits = "release"
 debug = true
 
+[profile.release]
+strip = "symbols"
+
 [patch.crates-io]
 html5ever = { git = "https://github.com/devongovett/html5ever.git", branch = "xmlns-attrs" }
 xml5ever = { git = "https://github.com/devongovett/html5ever.git", branch = "xmlns-attrs" }

--- a/scripts/build-native.js
+++ b/scripts/build-native.js
@@ -40,6 +40,10 @@ async function build() {
         args.push('--target', process.env.RUST_TARGET);
       }
 
+      if (process.env.ZIG_GLIBC) {
+        args.push('--zig', '--zig-abi-suffix', process.env.ZIG_GLIBC);
+      }
+
       let yarn = spawn('yarn', args, {
         stdio: 'inherit',
         cwd: pkg,

